[GoFr SOC] Improve Unit Test Coverage Across 6 Core Packages (zip, logger, pubsub/google, etc.)

Ashish-Kumar-Dash opened the pull request:
Description:

Added and enhanced unit tests to improve test coverage for the following packages:

pkg/gofr/file/zip.go

pkg/gofr/logging/logger.go

pkg/gofr/datasource/pubsub/google

pkg/gofr/migration/sql.go

pkg/gofr/datasource/sql/db.go

pkg/gofr/http/form_data_binder.go

Addressed edge cases, nil checks, and error handling branches that were previously untested.

Motivation: Boost reliability and confidence in core utilities and data layer by ensuring critical paths are covered.

Related to ongoing gofr SOC coverage enhancement initiative.

Coverage Summary:
File: file/zip.go — Before: 79.1% | After: 88.4% | Change: +9.3%

File: logging/logger.go — Before: 84.5% | After: 88.5% | Change: +4.0%

File: datasource/pubsub/google — Before: ~41.9% | After: 43.8% | Change: +1.9%

File: migration/sql.go — Before: 38.1% | After: 81.7% | Change: +43.6%

File: datasource/sql/db.go — Before: 55.6% | After: 91.2% | Change: +35.6%

File: http/form_data_binder.go — Added form_data_binder_test.go to cover core logic

Breaking Changes:

None introduced. All enhancements are test-only and non-breaking.

Additional Information:

Dependencies used:

pstest for in-memory Pub/Sub testing

testify for assertions

gomock for mocking interfaces

No production code changed. All changes are isolated to test files.

Coverage measured using go test -coverprofile and go tool cover -func.

Checklist:

[x] I have formatted my code using goimport and golangci-lint.
[x] All new code is covered by unit tests.
[x] This PR does not decrease the overall code coverage.
[x] I have reviewed the code comments and documentation for clarity.
@Umang01-hash  kindly review and advise on any changes, if needed
Thank You

Umang01-hash commented on pkg/gofr/datasource/pubsub/google/google_test.go:
```diff
@@ -268,98 +375,263 @@ func TestGoogleClient_CreateTopic_Error(t *testing.T) {
+
+}
+
+func TestGoogleClient_Query_EmptySubscription(t *testing.T) {
```
This test is failing with the following error message:

```go
=== RUN   TestGoogleClient_Query_EmptySubscription
    google_test.go:556: 
        	Error Trace:	...../gofr/pkg/gofr/datasource/pubsub/google/google_test.go:556
        	Error:      	An error is expected but got nil.
        	Test:       	TestGoogleClient_Query_EmptySubscription
--- FAIL: TestGoogleClient_Query_EmptySubscription (1.00s)
```

Umang01-hash commented on pkg/gofr/datasource/sql/db_test.go:
```diff
@@ -1142,3 +1142,71 @@ func TestClean(t *testing.T) {
 
 	assert.Empty(t, out)
 }
+func TestDB_PrepareContext(t *testing.T) {
```
This test is also failing with following error:
```go
=== RUN   TestDB_PrepareContext
    controller.go:97: missing call(s) to *sql.MockMetrics.RecordHistogram(is anything, is equal to app_sql_stats (string), is anything, is equal to hostname (string), is anything, is equal to database (string), is anything, is equal to type (string), is equal to SELECT (string)) ...../gofr/pkg/gofr/datasource/sql/db_test.go:1154
    controller.go:97: aborting test due to missing call(s)
--- FAIL: TestDB_PrepareContext (0.00s)
```

Umang01-hash commented on pkg/gofr/datasource/sql/db_test.go:
```diff
@@ -1142,3 +1142,71 @@ func TestClean(t *testing.T) {
+	err := db.Ping()
+	assert.Equal(t, sql.ErrConnDone, err)
+}
+func TestDB_ExecContextCancelled(t *testing.T) {
```
There is a panic when running this test, please resolve it:
```go
=== RUN   TestDB_ExecContextCancelled
&{ExecContext INSERT INTO dummy VALUES(1) 0 []}
--- FAIL: TestDB_ExecContextCancelled (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x18 pc=0x100ac28f8]

goroutine 174 [running]:
testing.tRunner.func1.2({0x100c5cf40, 0x101070db0})
	/usr/local/go/src/testing/testing.go:1734 +0x1ac
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1737 +0x334
panic({0x100c5cf40?, 0x101070db0?})
	/usr/local/go/src/runtime/panic.go:792 +0x124
gofr.dev/pkg/gofr/datasource/sql.(*DB).sendOperationStats(0x140004ac6c0, {0x1007cb840?, 0x1400018bd60?, 0x101084f20?}, {0x100b35dff, 0xb}, {0x100b3e4af, 0x1b}, {0x0, 0x0, ...})
	.../pkg/gofr/datasource/sql/db.go:57 +0x2b8
gofr.dev/pkg/gofr/datasource/sql.(*DB).ExecContext(0x140004ac6c0, {0x100cda500, 0x1400061fc70}, {0x100b3e4af, 0x1b}, {0x0, 0x0, 0x0})
	.../gofr/datasource/sql/db.go:99 +0x164
gofr.dev/pkg/gofr/datasource/sql.TestDB_ExecContextCancelled(0x140003dfa40)
	.../gofr/pkg/gofr/datasource/sql/db_test.go:1210 +0xb0
testing.tRunner(0x140003dfa40, 0x100cd1398)
	/usr/local/go/src/testing/testing.go:1792 +0xe4
created by testing.(*T).Run in goroutine 1
	/usr/local/go/src/testing/testing.go:1851 +0x374


```

Umang01-hash commented on pkg/gofr/http/form_data_binder_test.go:
```diff
@@ -0,0 +1,57 @@
+
+	})
+
+	t.Run("Unsupported Kind", func(t *testing.T) {
```
This is failing with following error:
```go
Test_setFieldValueFromData/Unsupported_Kind
    form_data_binder_test.go:55: 
        	Error Trace:	.../pkg/gofr/http/form_data_binder_test.go:55
        	Error:      	Received unexpected error:
        	            	unsupported type for field: , map[string]string
        	Test:       	Test_setFieldValueFromData/Unsupported_Kind
--- FAIL: Test_setFieldValueFromData/Unsupported_Kind (0.00s)
```

Umang01-hash commented on pkg/gofr/migration/sql_test.go:
```diff
@@ -106,160 +141,260 @@ func TestQueryRowContext(t *testing.T) {
+
+}
+
+func TestRollbackWithTransaction(t *testing.T) {
```
This test failed with following error:
```go
=== RUN   TestRollbackWithTransaction
    sql.go:113: Unexpected call to *migration.Mockmigrator.beginTransaction([&{0x14000848fc0   map[] 0x14000857b40 0x1400087c4e0 <nil> 0x1400087c3f0 0x140004f9ec0 0x1400087c408 0x1400087c420 0x1400087c438 <nil> 0x1400087c468 0x1400087c480 <nil> 0x1400087c4b0 0x1400087c498 0x1400087c4c8 0x1400087c450 0x14000857b20}]) at .../pkg/gofr/migration/sql.go:113 because: there are no expected calls of the method "beginTransaction" for that receiver
--- FAIL: TestRollbackWithTransaction (0.00s)
```

Umang01-hash reviewed (changes requested):
Hey @up1 Thankyou for your contribution. There are a few linter as well as test failures in your PR. Please fix them!

Ashish-Kumar-Dash commented:
hey @Umang01-hash  I have fixed the linting errors & the errors in the test cases.I have modified the logic of some test cases,and the ones I haven't been able to resolve have been replaced by new test cases that keep the updated level of coverage ~same  as shown in my PR description. Kindly review and let me know if any changes are needed,
Thank you

Ashish-Kumar-Dash commented:
hey @coolwednesday Regarding the lint issues,on first seeing the gci error i ran the gci tool for formatting but after i pushed new errors came up of extra white space and white lines,so, for past 2 commits i have run both golangci-lint run & golangci-lint run --output.text.print-issued-lines --output.text.colors=true --show-stats=false --timeout=5m(lint root module test command), and they have passed them locally,but always there is some hidden error Github Workflows finds which when i run the same test on my machine it doesn't catch it. It has been kind of frustrating, im attaching the ss of tests i ran of the most recent commit (fix lint issues) and you can see no error was flagged. 
<img width="1043" height="293" alt="Screenshot 2025-08-08 180708" src="https://github.com/user-attachments/assets/e2c6b897-ece1-4012-a65e-869ef3f20e5c" />
If Github Workflows catch a linting error now then I will be keen to hear what else I can do, which I haven't previously tried, because seeing through SS you can see golangci-lint caught no error.
Thank You

Umang01-hash reviewed (commented):
- Please remove the `go.work.sum` changes as it is not related to your PR. Else apart the PR looks fine and can be merged. And regarding the linter issues i just fixed it for you. Sometimes the version of linter and all creates these situations where in your local you are not getting any warnings but on github pipeline it shows linter errors. I think one thing you must make sure that linter versions in your local as well as one used in workflow are same.

Ashish-Kumar-Dash commented:
> hey @coolwednesday @Umang01-hash Regarding the lint issues,on first seeing the gci error i ran the gci tool for formatting but after i pushed new errors came up of extra white space and white lines,so, for past 2 commits i have run both golangci-lint run & golangci-lint run --output.text.print-issued-lines --output.text.colors=true --show-stats=false --timeout=5m(lint root module test command), and they have passed them locally,but always there is some hidden error Github Workflows finds which when i run the same test on my machine it doesn't catch it. It has been kind of frustrating, im attaching the ss of tests i ran of the most recent commit (fix lint issues) and you can see no error was flagged. <img alt="Screenshot 2025-08-08 180708" width="1043" height="293" src="https://private-user-images.githubusercontent.com/186616846/476019620-e2c6b897-ece1-4012-a65e-869ef3f20e5c.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTUwMDUwNDcsIm5iZiI6MTc1NTAwNDc0NywicGF0aCI6Ii8xODY2MTY4NDYvNDc2MDE5NjIwLWUyYzZiODk3LWVjZTEtNDAxMi1hNjVlLTg2OWVmM2YyMGU1Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwODEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDgxMlQxMzE5MDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iMTRmZDQyYmNlMTU3ZjA5NGI2ZTcxOTkyZThkMzc3M2M5MDM2NWU3Yzk4OTQwYWQ2YzllYmYwOGRmMTU3Mjk5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.JOrSk4foT_gT-BaXZ8IptggDd5TitCZ2357aiGCGUsA"> If Github Workflows catch a linting error now again then I will be keen to hear what else I can do, which I haven't previously tried, because seeing through SS you can see golangci-lint caught no error. Thank You

hey @Umang01-hash  @coolwednesday  I'm very grateful you corrected and reviewed my commits. Regarding the lint errors, I can't show you what I was talking about here, but its a thing of the past now and I have learned from my errors. I apologize if my comment was too "strong" . This has been a unique experience and I am very grateful for your prompt response. This PR has been a steep learning curve for me in my young  and I'm happy that my commits have passed the lint check. Kindly inform if there still any edits that need to be made .
Thank You,
Ashish Kumar Dash, IIT Bhilai CSE 28'

Umang01-hash commented on pkg/gofr/datasource/sql/db_test.go:
```diff
@@ -1142,3 +1143,86 @@ func TestClean(t *testing.T) {
+	err := db.Ping()
+	assert.Equal(t, sql.ErrConnDone, err)
+}
+func TestDB_QueryRow_NoRows(t *testing.T) {
```
<img width="1064" height="509" alt="Screenshot 2025-08-13 at 11 51 34 AM" src="https://github.com/user-attachments/assets/c61af43e-7854-4f82-8a58-8067ebd1915f" />

This test is panicing While running please fix it.

Ashish-Kumar-Dash commented:
fixed test cases across modules, ran locally and tested, edited /replaced tests which were panicking. Kindly run the workflow check  & review @Umang01-hash
